walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,179 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Collections;
+import java.util.HashSet;
 
+/**
+* FindMeetingQuery Class - Uses Event, Meeting Request, and Time Range objects
+* to schedule a meeting time given participants and unique schedules.
+*
+* A query is fielded in the following methodology:
+*      Collect busy times
+*      Calculate free windows
+*      Choose best window
+*      Figure which time works best
+*   
+*     If optional attendees
+*      Collect optional busy times
+*      Calculate those free windows
+*      Calculate intersection
+*/
 public final class FindMeetingQuery {
+
+  /**
+  * Function query
+  * This method searches for common free windows in the schedules of the attendees needed. 
+  * Optional attendees for the meeting are also considered and accommodated if possible.
+  * @param request The object containing the requesting event and details
+  * @param events A list of all of the events and attendees for the day
+  * @return A list of possible time frames that satisfy the request.
+  *
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    ArrayList<TimeRange> candidates = new ArrayList<>();
+    ArrayList<String> guests = new ArrayList<>(request.getAttendees());
+    ArrayList<String> optionals = new ArrayList<>(request.getOptionalAttendees());
+    long duration = request.getDuration();
+    
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    } else if (events.isEmpty() || guests.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    if (!optionals.isEmpty()) {
```

Comment:
nit:  you can have the empty case first. That's easier to read.

Reply:
Addressed: 00853d2

Reply:
Addressed: 1d79835